File: app/src/test/java/com/megablok10/app/qr/Mb10QrCodecTest.kt
package com.megablok10.app.qr

import com.megablok10.app.breach.Container
import com.megablok10.app.breach.LootSlot
import com.megablok10.app.breach.LootType
import com.megablok10.app.breach.Tier
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

/**
 * Свободный текст (позывной, имя точки, текст шарда) идёт в QR через base64
 * именно чтобы двоеточия/пробелы/кириллица внутри него не ломали разбор —
 * эти тесты специально гоняют такой "неудобный" текст через полный цикл
 * encode -> decode, а не только простые ASCII-строки без пунктуации.
 */
class Mb10QrCodecTest {

    @Test
    fun `contact round-trips`() {
        val raw = Mb10QrCodec.encodeContact("pubKeyB64==", "Zipa_09", "Otryad_SB")
        val decoded = Mb10QrCodec.decode(raw)
        assertEquals(Mb10Qr.Contact("pubKeyB64==", "Zipa_09", "Otryad_SB"), decoded)
    }

    @Test
    fun `container round-trips with punctuation in the name and a loot slot`() {
        val container = Container(
            id = "container-vent-8",
            name = "Панель вентиляции: техэтаж, блок Б",
            tier = Tier.HARD,
            ownerFaction = "Otryad_SB",
            loot = listOf(LootSlot(type = LootType.SHARD, tier = Tier.BASE, copies = 3, payload = "cGF5bG9hZA=="))
        )
        val raw = Mb10QrCodec.encodeContainer(container)
        val decoded = Mb10QrCodec.decode(raw)
        assertEquals(Mb10Qr.ContainerQr(container), decoded)
    }

    @Test
    fun `legacy AP QR still decodes as a tier-BASE container with no loot`() {
        // Формат до ревизии v9 (MB10:AP:v1:id:b64(name)) — уже напечатанные до неё QR обязаны продолжать читаться.
        val decoded = Mb10QrCodec.decode("MB10:AP:v1:ap-vent-8:0J/QsNC90LXQu9GM")
        assertEquals(
            Mb10Qr.ContainerQr(Container(id = "ap-vent-8", name = "Панель", tier = Tier.BASE, ownerFaction = "", loot = emptyList())),
            decoded
        )
    }

    @Test
    fun `shard round-trips with colons, newlines and cyrillic in every free-text field`() {
        val raw = Mb10QrCodec.encodeShard(
            id = "shard-report-1",
            decryptAction = true,
            tier = 2,
            valueHint = "ценный тех. документ",
            title = "Служебный лог: клиника",
            meta = "получен 21:02 · клиника, уровень доступа 2",
            body = "«...партия С-9: брак клапана, ответа не было...»"
        )
        val decoded = Mb10QrCodec.decode(raw)
        assertEquals(
            Mb10Qr.Shard(
                id = "shard-report-1",
                decryptAction = true,
                tier = 2,
                valueHint = "ценный тех. документ",
                title = "Служебный лог: клиника",
                meta = "получен 21:02 · клиника, уровень доступа 2",
                body = "«...партия С-9: брак клапана, ответа не было...»"
            ),
            decoded
        )
    }

    @Test
    fun `shard decryptAction flag survives round-trip both ways`() {
        val lockedRaw = Mb10QrCodec.encodeShard("s1", decryptAction = true, tier = 1, valueHint = "", title = "t", meta = "m", body = "b")
        val publicRaw = Mb10QrCodec.encodeShard("s2", decryptAction = false, tier = 1, valueHint = "", title = "t", meta = "m", body = "b")
        assertEquals(true, (Mb10QrCodec.decode(lockedRaw) as Mb10Qr.Shard).decryptAction)
        assertEquals(false, (Mb10QrCodec.decode(publicRaw) as Mb10Qr.Shard).decryptAction)
    }

    @Test
    fun `transaction round-trips with memo containing colons`() {
        val tx = Mb10Qr.Transaction(
            id = "tx-1",
            fromPubKeyB64 = "pubKeyB64==",
            amount = 350,
            memo = "Антидот: у контрабандиста",
            signatureB64 = "sigB64=="
        )
        val decoded = Mb10QrCodec.decode(Mb10QrCodec.encodeTransaction(tx))
        assertEquals(tx, decoded)
    }

    @Test
    fun `transaction signature payload is deterministic for the same inputs`() {
        val a = Mb10QrCodec.transactionSignaturePayload("tx-1", "pub==", 100, "за информацию")
        val b = Mb10QrCodec.transactionSignaturePayload("tx-1", "pub==", 100, "за информацию")
        assertEquals(String(a), String(b))
    }

    @Test
    fun `transaction signature payload changes if amount is tampered`() {
        val original = Mb10QrCodec.transactionSignaturePayload("tx-1", "pub==", 100, "оплата")
        val tampered = Mb10QrCodec.transactionSignaturePayload("tx-1", "pub==", 100000, "оплата")
        assert(!original.contentEquals(tampered))
    }

    @Test
    fun `receipt round-trips`() {
        val receipt = Mb10Qr.Receipt(id = "tx-1", receiverPubKeyB64 = "recvPub==", signatureB64 = "sigB64==")
        val decoded = Mb10QrCodec.decode(Mb10QrCodec.encodeReceipt(receipt))
        assertEquals(receipt, decoded)
    }

    @Test
    fun `receipt signature payload changes if receiver key is tampered`() {
        val original = Mb10QrCodec.receiptSignaturePayload("tx-1", "recvPub==")
        val tampered = Mb10QrCodec.receiptSignaturePayload("tx-1", "otherPub==")
        assert(!original.contentEquals(tampered))
    }

    @Test
    fun `unknown magic prefix decodes to null`() {
        assertNull(Mb10QrCodec.decode("NOTMB10:CONTACT:v1:a:b:c"))
    }

    @Test
    fun `unknown type segment decodes to null`() {
        assertNull(Mb10QrCodec.decode("MB10:UNKNOWN:v1:a:b:c"))
    }

    @Test
    fun `truncated transaction decodes to null instead of throwing`() {
        assertNull(Mb10QrCodec.decode("MB10:TX:v1:only-id"))
    }

    @Test
    fun `non-numeric amount decodes to null instead of throwing`() {
        assertNull(Mb10QrCodec.decode("MB10:TX:v1:tx-1:pub==:notanumber:bWVtbw==:sig=="))
    }

    @Test
    fun `garbage base64 in a free-text field decodes to null instead of throwing`() {
        assertNull(Mb10QrCodec.decode("MB10:AP:v1:ap-1:not-valid-base64!!!"))
    }
}

File: admin-web/server/src/mb10QrCodec.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { encodeContainerQr, encodeRamUpgradeQr, encodeShardQr, type ContainerSlotForQr } from "./lib/mb10QrCodec.js";

/**
 * mb10QrCodec.ts кодирует ТОЛЬКО (декодирует QR исключительно телефон
 * игрока — см. комментарий в самом файле), поэтому "round-trip" здесь
 * означает: разобрать строку тем же алгоритмом, что и decode() в
 * app/src/main/java/com/megablok10/app/qr/Mb10Qr.kt (Mb10QrCodec.decode),
 * скопированным сюда как эталон, и убедиться, что то, что генерирует
 * сервер, парсится клиентом байт-в-байт так же, как задумано. Если формат
 * разъедется — эти тесты должны сломаться раньше, чем реальный QR на игре.
 */
const MAGIC = "MB10";

function unb64(text: string): string {
  return Buffer.from(text, "base64").toString("utf8");
}

function decodeContainer(parts: string[]) {
  if (parts.length < 8) return null;
  const lootBlob = unb64(parts[7]);
  const loot =
    lootBlob.length === 0
      ? []
      : lootBlob.split(";").map((slot) => {
          const f = slot.split(",");
          const payload = f.slice(3).join(","); // limit=4 на Kotlin-стороне — всё после 3-й запятой это payload
          return { type: f[0], tierLevel: Number(f[1]), copies: Number(f[2]), payload };
        });
  return {
    id: parts[3],
    name: unb64(parts[4]),
    tierLevel: Number(parts[5]),
    ownerFaction: unb64(parts[6]),
    loot,
  };
}

function decodeShard(parts: string[]) {
  if (parts.length < 10) return null;
  return {
    id: parts[3],
    decryptAction: parts[4] === "1",
    tier: Number(parts[5]),
    valueHint: unb64(parts[6]),
    title: unb64(parts[7]),
    meta: unb64(parts[8]),
    body: unb64(parts[9]),
    moneyAmount: parts[10] !== undefined ? Number(parts[10]) : 0,
  };
}

function decodeRamUpgrade(parts: string[]) {
  if (parts.length < 5) return null;
  return { token: parts[3], delta: Number(parts[4]) };
}

function decode(raw: string) {
  const parts = raw.split(":");
  if (parts.length < 2 || parts[0] !== MAGIC) return null;
  switch (parts[1]) {
    case "CONTAINER":
      return { kind: "CONTAINER" as const, value: decodeContainer(parts) };
    case "SHARD":
      return { kind: "SHARD" as const, value: decodeShard(parts) };
    case "RAM":
      return { kind: "RAM" as const, value: decodeRamUpgrade(parts) };
    default:
      return null;
  }
}

test("encodeContainerQr — раунд-трип с несколькими слотами (SHARD + DAEMON)", () => {
  const slots: ContainerSlotForQr[] = [
    { typeName: "SHARD", tierLevel: 2, copies: 1, payload: "cGF5bG9hZC1vbmU=" },
    { typeName: "DAEMON", tierLevel: 1, copies: 0, payload: "cGF5bG9hZC10d28=" },
  ];
  const qr = encodeContainerQr("nasos-4", "Насосная-4", 2, "NEON_DRAGONS", slots);

  const decoded = decode(qr);
  assert.equal(decoded?.kind, "CONTAINER");
  const container = decoded!.value!;
  assert.equal(container.id, "nasos-4");
  assert.equal(container.name, "Насосная-4");
  assert.equal(container.tierLevel, 2);
  assert.equal(container.ownerFaction, "NEON_DRAGONS");
  assert.equal(container.loot.length, 2);
  assert.deepEqual(container.loot[0], { type: "SHARD", tierLevel: 2, copies: 1, payload: "cGF5bG9hZC1vbmU=" });
  assert.deepEqual(container.loot[1], { type: "DAEMON", tierLevel: 1, copies: 0, payload: "cGF5bG9hZC10d28=" });
});

test("encodeContainerQr — пустой список слотов декодируется в пустой лут, не в массив с одним мусорным элементом", () => {
  const qr = encodeContainerQr("empty-box", "Пустой ящик", 1, "", []);
  const decoded = decode(qr);
  assert.equal(decoded?.kind, "CONTAINER");
  assert.deepEqual(decoded!.value!.loot, []);
});

test("encodeShardQr — раунд-трип, decryptAction=true и moneyAmount сохраняются", () => {
  const qr = encodeShardQr("shard-1", true, 3, "дорогой на вид", "Логи СБ", "meta-x", "СЕКРЕТНЫЙ ТЕКСТ", 150);
  const decoded = decode(qr);
  assert.equal(decoded?.kind, "SHARD");
  assert.deepEqual(decoded!.value, {
    id: "shard-1",
    decryptAction: true,
    tier: 3,
    valueHint: "дорогой на вид",
    title: "Логи СБ",
    meta: "meta-x",
    body: "СЕКРЕТНЫЙ ТЕКСТ",
    moneyAmount: 150,
  });
});

test("encodeShardQr — decryptAction=false и moneyAmount=0 (значения по умолчанию)", () => {
  const qr = encodeShardQr("shard-2", false, 1, "", "Пустышка", "", "тело", 0);
  const decoded = decode(qr);
  assert.equal(decoded?.kind, "SHARD");
  assert.equal(decoded!.value!.decryptAction, false);
  assert.equal(decoded!.value!.moneyAmount, 0);
});

test("encodeRamUpgradeQr — раунд-трип", () => {
  const qr = encodeRamUpgradeQr("ram-abc123", 2);
  const decoded = decode(qr);
  assert.deepEqual(decoded, { kind: "RAM", value: { token: "ram-abc123", delta: 2 } });
});

test("decode — отклоняет строку без правильного MAGIC-префикса", () => {
  const notOurs = "OTHER:CONTAINER:v1:x";
  assert.equal(decode(notOurs), null);
});

test("decode — отклоняет усечённый CONTAINER (не долетели все сегменты)", () => {
  const qr = encodeContainerQr("nasos-4", "Насосная-4", 2, "NEON_DRAGONS", []);
  const truncated = qr.split(":").slice(0, 5).join(":"); // обрезаем последние сегменты
  const decoded = decode(truncated);
  assert.equal(decoded?.kind, "CONTAINER");
  assert.equal(decoded!.value, null);
});

test("encodeContainerQr — свободный текст с ':' и ';' внутри имени/фракции не ломает разбор (всё через base64)", () => {
  const qr = encodeContainerQr("nasos-5", "Название: с двоеточием; и точкой с запятой", 1, "Фракция:X", []);
  const decoded = decode(qr);
  assert.equal(decoded?.kind, "CONTAINER");
  assert.equal(decoded!.value!.name, "Название: с двоеточием; и точкой с запятой");
  assert.equal(decoded!.value!.ownerFaction, "Фракция:X");
});
